src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -724,14 +692,92 @@ internal override SyntaxNode ForbiddenZone
+            public LocalSymbolWithEnclosingContext(
+                Symbol containingSymbol,
+                Binder scopeBinder,
+                Binder nodeBinder,
+                TypeSyntax typeSyntax,
+                SyntaxToken identifierToken,
+                LocalDeclarationKind declarationKind,
+                SyntaxNode nodeToBind,
+                SyntaxNode forbiddenZone)
+                : base(containingSymbol, scopeBinder, false, typeSyntax, identifierToken, declarationKind)
+            {
+                Debug.Assert(
+                    nodeToBind.Kind() == SyntaxKind.CasePatternSwitchLabel ||
+                    nodeToBind.Kind() == SyntaxKind.ArgumentList && nodeToBind.Parent is ConstructorInitializerSyntax ||
+                    nodeToBind.Kind() == SyntaxKind.VariableDeclarator ||
+                    nodeToBind is ExpressionSyntax);
+                this._nodeBinder = nodeBinder;
+                this._nodeToBind = nodeToBind;
+                this._forbiddenZone = forbiddenZone;
+            }
+
+            internal override SyntaxNode ForbiddenZone => _forbiddenZone;
+
+            // This type is currently used for out variables and pattern variables.
+            // Pattern variables do not have a forbidden zone, so we only need to produce
+            // the diagnostic for out variables here.
+            internal override ErrorCode ForbiddenDiagnostic => ErrorCode.ERR_ImplicitlyTypedOutVariableUsedInTheSameArgumentList;
+
+            protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+            {
+                switch (_nodeToBind.Kind())
+                {
+                    case SyntaxKind.ArgumentList:
+                        var invocation = (ConstructorInitializerSyntax)_nodeToBind.Parent;
+                        _nodeBinder.BindConstructorInitializer(invocation.ArgumentList, (MethodSymbol)ContainingSymbol, diagnostics);
+                        break;
+                    case SyntaxKind.CasePatternSwitchLabel:
+                        _nodeBinder.BindPatternSwitchLabelForInference((CasePatternSwitchLabelSyntax)_nodeToBind, diagnostics);
+                        break;
+                    case SyntaxKind.VariableDeclarator:
+                        // This occurs, for example, in
+                        // int x, y[out var Z, 1 is int I];
+                        // for (int x, y[out var Z, 1 is int I]; ;) {}
+                        var declarator = (VariableDeclaratorSyntax)_nodeToBind;
+                        if (declarator.ArgumentList != null)
+                        {
+                            foreach (var arg in declarator.ArgumentList.Arguments)
+                            {
+                                var expression = arg.Expression;
+                                if (expression.Kind() == SyntaxKind.DeclarationExpression)
+                                {
+                                    this.SetType(_nodeBinder.CreateErrorType("var"));
```

Comment:
>this.SetType(_nodeBinder.CreateErrorType("var")); [](start = 36, length = 49)

It is not clear why are we making assumption that this expression is related to the local symbol. I believe there could be scenarios where inference for some locals can succeed even when some arguments are declaration expressions. It feels like we should simply be calling BindDeclaratorArguments instead of looping through the arguments here.